Add "Today" to the Date Picker

HardeepAsrani opened the pull request:
## Description
Added "Today" button to React Date Picker in the post-schedule box, which fixes this: #6881

It also has a `utcOffset` parameter to make sure the "Today" is according to WordPress' timezone, not browser's timezone.

Only added it to the Post Schedule as the button might not be required in all the cases.

## How has this been tested?
Yes, tested with the Gutenberg tools and linters to make sure everything is fine, and tried on various browsers and timezones to make it work.


## Types of changes
Enhancement to Date Picker in Post Schedule.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

HardeepAsrani committed:
Merge pull request #1 from WordPress/master

Update

HardeepAsrani committed:
Merge pull request #2 from WordPress/master

Merge changes

HardeepAsrani committed:
Merge pull request #3 from WordPress/master

Fix CodeEditor not loading when WordPress is installed in a subfolder…

HardeepAsrani committed:
Add "Today" to the Date Picker

Added "Today" button to React Date Picker in Scheduled box, which fixes this: https://github.com/WordPress/gutenberg/issues/6881

It also has a `utcOffset` parameter to make sure the "Today" is according to WordPress' timezone, not browser's timezone.

youknowriad commented on editor/components/post-schedule/index.js:
```diff
@@ -25,6 +26,8 @@ export function PostSchedule( { date, onUpdateDate } ) {
 			key="date-time-picker"
 			currentDate={ date }
+			utcOffset={ settings.timezone.offset }
+			todayButton={ __( 'Today' ) }
```
Do we want this in all DatePicker components? in which case it's probably better to add it to the DatePicker component instead.

Also, why do we need the `utcOffset` here?

HardeepAsrani replied:
I don't think it needs to be in all the date pickers. Some might use to to select date of birth or other uses, and won't be needed there. 

We need ofset to make sure the "Today" it chooses is according to the timezone set in WordPress settings.

I'm in India and my local setup has GMT 0 timezone (23rd of May), so when I clicked on Today around 1am IST time (24th of May), and WordPress took it as a future"Scheduled" post.

HardeepAsrani replied:
I realize my explanation was really confusing. Basically to make sure "Today" is taken from the timezone defined in WP settings, not the browser...which will cause the posts to be published in past or future if you're not from the same timezone as your WP. Hope this makes more sense. 😁

youknowriad replied:
Thanks for the clarification, it makes sense but do you think we should avoid using the React Date Picker props and think of better prop names/values because the underlying implementation can change later.

I think `utcOffset` make sense but instead of `todayButton` I'd prefer a boolean only prop. `showTodayButton`. And let's document these props in the README of the `DateTimePicker`  component

HardeepAsrani replied:
I was thinking of doing the same thing but seemed like an overkill to simply things. But didn’t think of the possibility of react-date-picker ever changing their implementation. I’ll update the PR tonight. :)

youknowriad replied:
Actually, I was thinking of us moving away from react-date-picker ;) #1311

HardeepAsrani replied:
It says it is accessible but doesn't look like it at first glance at all. And you've confirmed that it's not that assessible. XD

I think the best way will be to fork it and make it accessible.

HardeepAsrani replied:
Wanted to ask you the best way to implement the solution you proposed in JSX. I've added `showTodayButton` to `<DateTimePicker>` element, and checked the condition in `ReactDatePicker` with the following:

```
todayButton={ showTodayButton ? __( 'Today' ) : undefined }
utcOffset={ showTodayButton ? settings.timezone.offset : undefined }
```

Or should I do it using a spreaded object instead?

youknowriad replied:
i'm fine with that :) but I think instead of accessing the date settings in the component I'd just pass an `utcOffset` prop like you did originally to avoid having to add a dependency to the `date` module in the `components` module.

HardeepAsrani replied:
Let me know if these changes are good. :)

HardeepAsrani committed:
Added `showTodayButton` prop to DateTimePicker

Added `showTodayButton` prop to pass boolean to display Today button in ReactDatePicker, as discusses at https://github.com/WordPress/gutenberg/pull/6927

youknowriad commented on components/date-time/README.md:
```diff
@@ -53,6 +55,13 @@ The function called when a new date or time has been selected. It is passed the
 
+### utcOffset
+
+The offset of current timezone.
```
Can we add a message saying that it's required when `shoowTodayButton` is true ?

HardeepAsrani replied:
Should I add, “Needed to be true if ‘showTodayButton’ is used”

youknowriad replied:
More like `The offset of current timezone.. This prop is required if `showTodayButton` is true.`

HardeepAsrani replied:
Done. Thanks. :)

youknowriad commented:
Code wise, this is good, waiting for a design review.

HardeepAsrani committed:
Improved utcOffset info

Added more details to utcOffset info in documentation

HardeepAsrani commented:
@jasmussen Why don't we fork the React Date Picker for Gutenberg and make it accessible, like WooCommerce did for select2

HardeepAsrani commented:
Not sure why the build failed as it was just text edit. 😄

jasmussen commented:
The ticket to follow is this one: #1311 —  so it's something we have to solve one way or another.

HardeepAsrani commented:
Let me know if there's anything else needed here. :)

tofumatt commented:
Right now there is a lot of talk about redesigning the date picker entirely, adding accessibility features, etc. (see #7621 and especially https://github.com/WordPress/gutenberg/pull/7621#issuecomment-401605022).

Because of that I'm gonna close this PR for now. In #7621 there was a lot of talk about not shipping half-baked stuff (the functionality here works but it needs design work) and this component is undergoing a lot of work in other discussions.

Thanks for your work on this; we can re-open it after #1311 gets addressed if it's still applicable to the date-picker component we're using and it needs addressing. 👍
